### PR TITLE
Concurrency cancellation

### PR DESCRIPTION
**Pull request by aron**
Currently a cancellation event sent to a worker with `concurrency` > 1 will always cancel the most recent task.

For the async `_aloop` we now use a weak map to track all in flight prediction tasks. On receipt of a `Cancel` event, we retrieve the task from the map and cancel it.

For the standard `_loop` method we continue to ignore the `Cancel` event. Relying on calling the `send_cancel_signal()` to terminate the worker method before sending the `Cancel` event. I'm not 100% clear on why the parent needs to do this but calling `send_cancel_signal` or raising `CancelationException` directly in the `_loop` method causes the tests to fail so I've left it as-is.

### CODE REVIEW

File: python/cog/server/worker.py [hunk 1]
```diff
@@ -516,7 +521,8 @@ def _loop(
         while True:
             e = cast(Envelope, self._events.recv())
             if isinstance(e.event, Cancel):
-                continue  # Ignored in sync predictors.
+                # this is handled via SIGUSR1 signals from the parent via send_cancel_signal
```
Comment (philandstuff) [inline python/cog/server/worker.py:524]:
```suggestion
                # for sync predictors, this is handled via SIGUSR1 signals from the parent via send_cancel_signal
```

File: python/cog/server/worker.py [hunk 2]
```diff
@@ -533,17 +539,27 @@ async def _aloop(
...
                 e = cast(Envelope, await self._events.recv())
-                if isinstance(e.event, Cancel) and task and self._cancelable:
+                if isinstance(e.event, Cancel):
+                    if not self._cancelable:
+                        continue
+
+                    task = tasks.get(e.tag)
+                    if not task:
+                        print(
+                            f"Got cancel event for unrecognized prediction",
```
Comment (philandstuff) [inline python/cog/server/worker.py:553]:
```suggestion
                            "Got cancel event for unrecognized prediction",
```

File: python/cog/server/worker.py [hunk 3]
```diff
@@ -533,17 +539,27 @@ async def _aloop(
...
 
-        task = None
-
         async with asyncio.TaskGroup() as tg:
+            tasks = weakref.WeakValueDictionary[str | None, asyncio.Task[Any]]()
             while True:
                 e = cast(Envelope, await self._events.recv())
-                if isinstance(e.event, Cancel) and task and self._cancelable:
+                if isinstance(e.event, Cancel):
+                    if not self._cancelable:
```
Comment (philandstuff) [inline python/cog/server/worker.py:547]: this isn't a useful flag for concurrent predictions (it's set and cleared in _handle_predict_error; but in the concurrent context it's set whenever *any* prediction starts and cleared whenever *any* prediction finishes); but I think we can rely entirely on whether the prediction is in the weak value dictionary.

**Comment by aron** [comment]
Hmm, I've broken the sync tests somewhere, putting this back into draft while I figure it out.

**Comment by aron** [comment]
I had an accidental `setup=False` in a WorkerConfig of the failing test... fixed now.

**Pull request merged**